02 - Variables, types, control flow.ipynb

```diff
@@ -0,0 +1,2206 @@
+      "\u001b[0;31mKeyError\u001b[0m: 'local_one'"
+     ]
+    }
+   ],
+   "source": [
+    "global_one = 1\n",
+    "\n",
+    "def foo():\n",
+    "    local_one = 2\n",
+    "    print(locals())\n",
+    "\n",
+    "\n",
+    "foo()\n",
+    "print(globals()[\"global_one\"])  # `print(globals())` will not have a very pretty output in the Jupyter notebook but you can try it\n",
+    "print(globals()[\"local_one\"])\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Всеки блок от код си има своя област на видимост, в която стоят локално дефинираните имена.\n",
+    "Ако една функция не може да намери дадена променлива в локалния си scope, търси в обграждащия (глобалния) за променлива със същото име:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 61,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1\n"
+     ]
+    }
+   ],
+   "source": [
+    "global_one = 1\n",
+    "\n",
+    "def foo():\n",
+    "    print(global_one)\n",
+    "\n",
+    "foo()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "По подразбиране пренасочването на имена става в локалния scope.\n",
+    "Използването на ключовата дума `global` позволява пренасочването на глобални имена. Това обаче въобще не е добра практика."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 62,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "2\n",
+      "{'global_one': 2}\n",
+      "1\n"
+     ]
+    }
+   ],
+   "source": [
+    "global_one = 1\n",
+    "\n",
+    "def foo():\n",
+    "    global_one = 2\n",
+    "    print(global_one)\n",
+    "    print(locals())\n",
+    "\n",
+    "foo()\n",
+    "\n",
+    "print(globals()[\"global_one\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Аргументите на функциите отиват в `locals` речника, достъпен от тялото им (съответно не и извън него):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 63,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'x': 1, 'y': 2, 'args': (3, 4, 5, 6, 7, 8, 9, 10), 'kwargs': {'name': 'Pesho', 'age': 20}}\n"
+     ]
+    }
+   ],
+   "source": [
+    "def fn(x, y, *args, **kwargs):\n",
+    "    print(locals())\n",
+    "\n",
+    "fn(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, name=\"Pesho\", age=20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "В случай, че имаме вложени блокове, ключовата дума `nonlocal` позволява пренасочване на име, дефинирано в обграждащ блок (за разлика от `global`, където името е дефинирано в глобалния обхват). Също както `global`, изполването му е лоша практика."
```

Comment:
Maybe we should give an example? https://www.programiz.com/python-programming/global-keyword
Also, we never mention what the keyword `global` does

Reply:
It is mentioned just a few cells above but there is no example. Maybe it won't hurt to give some examples yeah